libraries/AP_RTC/AP_RTC.cpp

```diff
@@ -0,0 +1,239 @@
+    uint32_t sec_ms = (time_ms % (60 * 1000)) - ms;
+    uint32_t min_ms = (time_ms % (60 * 60 * 1000)) - sec_ms - ms;
+    uint32_t hour_ms = (time_ms % (24 * 60 * 60 * 1000)) - min_ms - sec_ms - ms;
+
+    // convert times as milliseconds into appropriate units
+    sec = sec_ms / 1000;
+    min = min_ms / (60 * 1000);
+    hour = hour_ms / (60 * 60 * 1000);
+}
+
+bool AP_RTC::is_set() const
+{
+    for (uint8_t i=0; i<_num_rtc_sources; i++) {
+        const source_info &info = _source_info[i];
+        if (info.is_set) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// get milliseconds from now to a target time of day expressed as hour, min, sec, ms
+// match starts from first value that is not -1. I.e. specifying hour=-1, minutes=10 will ignore the hour and return time until 10 minutes past 12am (utc)
+uint32_t AP_RTC::get_time_utc(int32_t hour, int32_t min, int32_t sec, int32_t ms)
+{
+    // determine highest value specified (0=none, 1=ms, 2=sec, 3=min, 4=hour)
+    int8_t largest_element = 0;
+    if (hour != -1) {
+        largest_element = 4;
+    } else if (min != -1) {
+        largest_element = 3;
+    } else if (sec != -1) {
+        largest_element = 2;
+    } else if (ms != -1) {
+        largest_element = 1;
+    } else {
+        // exit immediately if no time specified
+        return 0;
+    }
+
+    // get start_time_ms as h, m, s, ms
+    int32_t curr_hour, curr_min, curr_sec, curr_ms;
+    get_system_clock_utc(curr_hour, curr_min, curr_sec, curr_ms);
+    int32_t total_delay_ms = 0;
+
+    // calculate ms to target
+    if (largest_element >= 1) {
+        total_delay_ms += ms - curr_ms;
+    }
+    if (largest_element == 1 && total_delay_ms < 0) {
+        return static_cast<uint32_t>(total_delay_ms += 1000);
+    }
+
+    // calculate sec to target
+    if (largest_element >= 2) {
+        total_delay_ms += (sec - curr_sec)*1000;
+    }
+    if (largest_element == 2 && total_delay_ms < 0) {
+        return static_cast<uint32_t>(total_delay_ms += (60*1000));
+    }
+
+    // calculate min to target
+    if (largest_element >= 3) {
+        total_delay_ms += (min - curr_min)*60*1000;
+    }
+    if (largest_element == 3 && total_delay_ms < 0) {
+        return static_cast<uint32_t>(total_delay_ms += (60*60*1000));
+    }
+
+    // calculate hours to target
+    if (largest_element >= 4) {
+        total_delay_ms += (hour - curr_hour)*60*60*1000;
+    }
+    if (largest_element == 4 && total_delay_ms < 0) {
+        return static_cast<uint32_t>(total_delay_ms += (24*60*60*1000));
+    }
+
+    // total delay in milliseconds
+    return static_cast<uint32_t>(total_delay_ms);
+}
+
+AP_RTC::source_id_t AP_RTC::register_rtc_source(AP_RTC::source_type type)
+{
+    if (_num_rtc_sources >= _max_rtc_sources) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "RTC: too many clock sources");
+        return _invalid_source;
+    }
+    gcs().send_text(MAV_SEVERITY_WARNING, "RTC: new %s clock source",
```

Comment:
I don't think we need this new clock source do we?  All is well so I don't think we should report to the GCS.